### PR TITLE
fix: harden C64 Bridge against 35 review findings (HARD01-001..035)

### PR DESCRIPTION
**Pull request by chrisgleissner**
## Summary

Implements every confirmed finding (**HARD01-001 through HARD01-035**) from the whole-server hardening review at [`doc/reviews/hardening/01-fable/review.md`](doc/reviews/hardening/01-fable/review.md), commissioned to find false-success, wrong-target, lifecycle, input, filesystem, SID, and packaging failures across the MCP server, Ultimate REST backend, VICE backend, and background tooling.

The public MCP contract is preserved **except** where an existing tool result was demonstrably false (e.g. `power_cycle` claiming success without verifying the highlighted menu item, `upload_run_asm` claiming execution when the code never ran). In those cases the tool now returns a clear, actionable failure instead.

### Ultimate REST truthfulness
- **HARD01-011**: Any C64U/U2 REST response with a non-empty firmware `errors[]` is now treated as failure, even on HTTP 200 — previously every `run_prg`/`mount`/`config`/etc. call reported success purely because the HTTP call didn't throw.
- **HARD01-002 / HARD01-003**: `power_cycle` now verifies the *highlighted* Tool Menu row via the colour matrix before sending the final RETURN, aborts if it can't confirm the selection, and attempts a best-effort menu close on any mid-flow failure.
- **HARD01-010**: Device filesystem paths are encoded per path segment so nested `/` separators stay literal instead of being double-encoded as `%2F`.
- **HARD01-023**: `drivesList()` is normalized to one `{id, power, image, type}` shape on both the c64u and VICE facades, so `drive_mount_and_verify` actually works against real firmware's `{errors, drives: [{id: info}]}` response.
- **HARD01-024**: `X-Password` (and other credential-shaped headers) are redacted case-insensitively before reaching debug logs/diagnostics.

### VICE lifecycle and transport
- **HARD01-012**: Binary Monitor requests now carry a per-request timeout that rejects, clears the pending-request map, and destroys the socket — one lost frame no longer wedges every later VICE operation forever.
- **HARD01-013**: SIGINT/SIGTERM/SIGHUP handlers now synchronously kill managed VICE/Xvfb processes (with a genuinely synchronous path usable from a process `exit` hook), so sessions don't leak orphaned processes.
- **HARD01-014**: VICE joystick input now uses the real Binary Monitor `0xA2` Joyport Set command instead of writing CIA data-port registers, which never actually simulated input.
- **HARD01-015**: Checkpoint operation masks honor exactly the load/store/execute flags requested instead of silently adding an execute trap.
- **HARD01-026**: VICE `reset()`/`reboot()` now returns `success:false` with readiness diagnostics when BASIC never reaches READY, instead of always reporting success.
- **HARD01-035**: VICE `configSet`/`configBatchUpdate` look up the resource's declared type before coercing, so a string resource whose value looks numeric (e.g. `"1541"`) is no longer silently turned into an integer.

### Backend-switch and config consistency
- **HARD01-016**: Multi-step flows (paused write-verify, a running SID arpeggio) now pin the facade they started on for their full lifetime via a new `pinFacade()`, so a concurrent `c64_select_backend` can no longer retarget an operation already underway.
- **HARD01-005 / HARD01-006 / HARD01-027**: One authoritative config loader with documented precedence (`C64BRIDGE_CONFIG` → cwd → home → legacy package-relative → defaults); a malformed config file now logs a redacted warning and falls back instead of crashing the server or silently disagreeing between two code paths.
- **HARD01-004**: A transient `machine:input` probe failure no longer poisons the capability cache as permanently "unknown" — it retries on the next read.
- **HARD01-025**: RAG/knowledge asset paths resolve from the installed package root, not `process.cwd()`, so an npm-installed server started from a project directory still finds its bundled data.
- **HARD01-031**: Reading `c64://platform/status` now races the live capability probe against a 750ms bound instead of blocking on a full device timeout.

### Program execution truthfulness
- **HARD01-018**: `upload_run_asm` now loads the assembled PRG (via a new `loadPrg` facade method) and enters it with a typed `SYS <entry>`, instead of LOAD+RUN — which misinterpreted raw 6510 opcodes as BASIC tokens and, for non-`$0801` origins, corrupted VICE's BASIC workspace pointers. This is the flagship fix: the documented assembly workflow now actually executes.
- **HARD01-019 / HARD01-020**: ASM liveness polling reads only screen RAM, never VIC/CIA/SID I/O — the old heuristic treated the free-running raster counter as "activity" (so it could never detect a crash) and, on real hardware, read the CIA interrupt-control registers, which acknowledges pending IRQs as a side effect.
- **HARD01-021**: BASIC error detection now requires the real `?` error prefix, so programs that print the word "ERROR" in ordinary output are no longer misreported as failed.
- **HARD01-007**: Printer text containing quotes is now spliced with `CHR$(34)`; the old doubled-quote (`""`) escaping is invalid Commodore BASIC V2 and threw a runtime `?SYNTAX ERROR`.
- **HARD01-008**: The generated PETSCII "wait for key" line actually loops until a key is pressed, instead of falling through to `READY.` immediately.
- **HARD01-009**: Keyboard queue injection now reports partial delivery via a typed timeout error instead of silently overwriting an undrained KERNAL queue.

### Background tasks, artifacts, and SID
- **HARD01-022**: Tasks persisted mid-"running" are downgraded to "stopped" with an explanatory `lastError` on reload, instead of becoming zombie entries that `list_tasks` reports as alive forever.
- **HARD01-028**: A background iteration that *resolves* `{success:false}` (rather than throwing) is now recorded as a failure.
- **HARD01-029**: `bundle_run_artifacts` validates `runId` and every derived filename, and proves containment via `path.relative` — a `runId` of `../escape` or an absolute path is rejected outright.
- **HARD01-030**: Generated SID arpeggios (`music_generate`) pin their facade at creation and are cancelled on backend switch / `sid_reset` / `sid_silence_all`, instead of an untracked detached loop that could keep writing notes to whatever backend became active later.
- **HARD01-032**: The SID frequency formula now uses the correct 24-bit phase-accumulator divisor (`hz * 2^24 / clock`) everywhere — code and knowledge docs alike. The old `2^16` formula made every generated note ~256x too low in pitch (sub-audible).
- **HARD01-033**: The envelope gate is now cleared between generated notes so ADSR actually retriggers, instead of attacking once and never articulating again.
- **HARD01-034**: Printer tools are now gated to `["c64u", "u2"]`, matching the platform-status capability that already advertised printer support on U2.

### Packaging
- **HARD01-001**: `--http [port]` now genuinely opens an HTTP MCP transport (`StreamableHTTPServerTransport`); default invocation is unchanged (stdio).

## Two regressions caught and fixed before this was ready

1. **VICE mock server didn't simulate KERNAL keyboard drain.** The HARD01-018 fix for real VICE timed out against the Binary Monitor mock (which doesn't run real 6502 code, so nothing drains the keyboard queue's pending-count byte). Added a short simulated drain to `src/vice/mockServer.ts`, matching the equivalent fix already needed for the c64u REST mock.
2. **An early HARD01-001 approach would have broken every real deployment.** A `process.argv[1] === import.meta.url` "am I the entrypoint" check seemed like a clean way to stop test imports from starting a live server — but the documented entrypoint (`src/index.ts` → `dist/index.js`, per `package.json` main/bin) always imports `mcp-server.ts` as a side effect rather than being invoked directly, so that check is always false in production and the server would never start. Replaced with an explicit `C64BRIDGE_SKIP_AUTO_START` opt-out env var, mirroring the existing `C64BRIDGE_START_SKIP_AUTO_LAUNCH` convention already used by `scripts/start.mjs`. Verified by actually running `node dist/index.js` and `node src/index.ts` end-to-end after the fix.

## Test plan

- [x] Every one of the 35 findings has a dedicated, named regression test — `grep -r "HARD01-" test/` to audit coverage.
- [x] `npx tsc -p tsconfig.json --noEmit` — clean.
- [x] `npm test` (full matrix, default c64u-mock target, both Node and Bun runners) — 0 failures.
- [x] `npm run test:vice:mock` (VICE Binary Monitor mock target) — 0 failures.
- [x] `npm run build` — succeeds; `mcp/*.json` and `README.md` regenerated and diff clean against source (generator output matches checked-in files).
- [x] `npm run check:package` — passes.
- [x] `git diff --check` — no whitespace issues.
- [x] Manual smoke test: `node dist/index.js` and `node --import tsx/esm src/index.ts` both start the server correctly end-to-end (read-only connectivity probe against a live device on the test network; no state-changing calls).
- [ ] Real hardware (C64U/U64/U2) or a real VICE binary — intentionally **not** exercised, per the task's constraint against state-changing operations on real devices. All coverage is unit/mock-server based.

## Notes for reviewers

- `doc/reviews/hardening/01-fable/` (the review itself and the prompts that produced it) is included for traceability between findings and fixes; it has not been edited to hide or soften any finding.
- `mcp/*.json` and `README.md` diffs are auto-generated (`npm run build`) from the source/schema changes above — no manual edits.

### CODE REVIEW

File: src/vice/process.ts [hunk 1]
```diff
@@ -404,11 +405,21 @@ export async function startViceProcess(options: ViceProcessOptions): Promise<Vic
     }
   };
 
+  const stopSync = (): void => {
+    for (const process of [child, xvfb]) {
+      if (!process || process.exitCode !== null || process.signalCode !== null) continue;
+      try { process.kill("SIGTERM"); } catch {}
```
[the remark this answers is not in the document]
Comment (chrisgleissner) [reply src/vice/process.ts]: Fixed in ad93c5d: `stopSync()` now sends `SIGKILL` directly instead of the dead-code SIGTERM-then-immediate-SIGKILL dance. This function is only ever called synchronously from signal handlers / the process `exit` hook, so there's no later tick to observe a graceful SIGTERM exit before escalating — going straight to SIGKILL is honest about what actually happens.

File: src/vice/viceClient.ts [hunk 2]
```diff
@@ -209,7 +215,7 @@ export class ViceClient {
     if (!this.socket || this.socket.destroyed) {
       return Promise.reject(new Error("VICE monitor socket is not connected"));
     }
-    const reqId = this.nextReqId++;
+    const reqId = this.nextReqId++ >>> 0;
```
[the remark this answers is not in the document]
Comment (chrisgleissner) [reply src/vice/viceClient.ts:221]: Fixed in ad93c5d, using your suggested guard: `nextReqId` now resets to 0 before incrementing when it reaches `0xFFFFFFFF`, so the reserved unsolicited-event marker is never issued as a request ID. Added a dedicated regression test ("ViceClient skips the reserved 0xFFFFFFFF reqId on wraparound...") that sets `nextReqId` to the boundary value and asserts the encoded request IDs wrap to 0 rather than colliding.

File: src/vice/mockServer.ts [hunk 3]
```diff
@@ -235,6 +235,14 @@ export class ViceMockServer {
         if (start <= SCREEN_BASE && end >= SCREEN_BASE) {
           this.helloReady = false;
         }
+        // This mock does not run real 6502 code, so nothing else drains the
+        // KERNAL keyboard buffer's pending-count byte ($00C6/NDX). Simulate
+        // the KERNAL IRQ consuming it shortly after each write, mirroring
+        // real VICE/hardware behaviour closely enough for deterministic tests.
+        const KEYBOARD_NDX = 0x00c6;
+        if (start <= KEYBOARD_NDX && end >= KEYBOARD_NDX) {
+          setTimeout(() => { this.memory[KEYBOARD_NDX] = 0; }, 5);
```
[the remark this answers is not in the document]
Comment (chrisgleissner) [reply src/vice/mockServer.ts]: Fixed in ad93c5d: added `.unref()` to the simulated NDX-drain timeout exactly as suggested.

File: src/mcp-server.ts [hunk 4]
```diff
@@ -520,11 +540,18 @@ function createPlatformResourceDescriptor() {
...
   const status = getPlatformStatus();
   const availableBackends = client.getAvailableBackends();
   const capabilities = describePlatformCapabilities(toolRegistry.list());
-  const nativeEndpoints = await client.getNativeEndpointCapabilities();
+  // Resource reads must remain responsive while an Ultimate boots or is
+  // unreachable. The probe itself remains retryable (unknown is not cached).
+  const nativeEndpoints = await Promise.race([
+    client.getNativeEndpointCapabilities(),
+    new Promise<Awaited<ReturnType<C64Client["getNativeEndpointCapabilities"]>>>((resolve) => {
+      setTimeout(() => resolve({ machineInput: "unknown", machineMenuScreen: "unknown" }), 750);
```
[the remark this answers is not in the document]
Comment (chrisgleissner) [reply src/mcp-server.ts]: Fixed in ad93c5d: added `.unref()` to the losing timer in the `Promise.race`, per your note that this is sufficient for an HTTP transport since `Promise.race` doesn't expose a handle to explicitly clear the loser.

File: test/device.test.mjs [hunk 5]
```diff
@@ -691,20 +753,40 @@ test("device: createFacade merges backend sections across config candidates", as
...
-      assert.equal(selected, "vice");
-      assert.equal(reason, "config only");
-      assert.equal(facade.type, "vice");
+      assert.match(reason, /fallback/);
+      assert.ok(selected === "c64u" || selected === "vice", `unexpected backend: ${selected}`);
+      assert.equal(facade.type, selected);
+      if (selected === "vice") {
+        // The real assertion this scenario exists for: home's vice section
+        // must not have leaked in via merging.
+        assert.notEqual(facade.exe, "/nonexistent/c64bridge-test-fixture/x64sc");
```
[the remark this answers is not in the document]
Comment (chrisgleissner) [reply test/device.test.mjs]: Fixed in ad93c5d: replaced the nonexistent fixture path with a real, uniquely-named temp executable created via `fs.mkdtempSync`/`writeFileSync`. Since the fixture actually exists on disk, `resolveViceBinary()`'s `findBinary()` would return it verbatim if home's config leaked through, so the `notEqual` assertion can now actually fail on a real leak — while a per-run mkdtemp path still can't coincidentally collide with the environment's real default-resolved VICE binary (the failure mode that caused the previous "/usr/bin/x64sc" fixture to be replaced with the now-broken impossible path in the first place). Fixture is cleaned up in a `finally` block.
